substate-store-impls/src/substate_database_staging.rs

```diff
@@ -0,0 +1,282 @@
+        // This function iterates over entries of the specified partition. Therefore, we don't need
+        // to think about other partitions here. We first check if there are any partition updates
+        // for the specified partition. If there is not, no overlaying is needed and we can just
+        // return the iterator of the root store.
+        match self.overlay.node_updates.get(node_key) {
+            // There is a partition update in the overlay.
+            Some(NodeDatabaseUpdates { partition_updates }) => {
+                match partition_updates.get(partition_num) {
+                    // The partition was reset. None of the substates of this partition that exist
+                    // in the root store "exist" anymore. We just need an iterator over the new
+                    // substates in the reset action.
+                    Some(PartitionDatabaseUpdates::Reset {
+                        new_substate_values,
+                    }) => Box::new(
+                        new_substate_values
+                            .iter()
+                            .map(|(sort_key, substate_value)| {
+                                (sort_key.clone(), substate_value.clone())
+                            })
+                            .sorted(),
+                    ),
+                    // There are some changes that need to be overlayed.
+                    Some(PartitionDatabaseUpdates::Delta { substate_updates }) => {
+                        let underlying = self.root.list_entries_from(partition_key, from_sort_key);
+                        let overlaying = substate_updates
+                            .iter()
+                            .map(|(sort_key, database_update)| match database_update {
+                                DatabaseUpdate::Set(substate_value) => {
+                                    (sort_key.clone(), Some(substate_value.clone()))
+                                }
+                                DatabaseUpdate::Delete => (sort_key.clone(), None),
+                            })
+                            .sorted_by(|(sort_key_a, _), (sort_key_b, _)| {
+                                sort_key_a.cmp(sort_key_b)
+                            });
+                        Box::new(OverlayingIterator::new(underlying, overlaying))
+                    }
+                    // Overlay doesn't contain anything for the provided partition number. Return an
+                    // iterator over the data in the root store.
+                    None => self.root.list_entries_from(partition_key, from_sort_key),
+                }
+            }
+            // Overlay doesn't contain anything for the provided node key. Return an iterator over
+            // the data in the root store.
+            None => self.root.list_entries_from(partition_key, from_sort_key),
+        }
+    }
+}
+
+impl<'s, S> CommittableSubstateDatabase for SubstateDatabaseStaging<'s, S> {
```

Comment:
What do you think about also implementing some kind of "commit into underlay" functionality (if S is committable?)

Reply:
I would slightly prefer to keep the database immutable since it makes it clear in the type system that the root database is never mutated in any way and that all that this type needs is read-only access to the database. If committing to the root database is needed, perhaps it can be done outside of the overlay. As an example, something like:

```rust
let mut root = InMemorySubstateDatabase::standard();
let overlay = SubstateDatabaseStaging::new(&root);
root.commit(&overlay.into_database_updates());
```

Reply:
Actually, there might be a very interesting way to achieve this in the Rust type system that allows us to operate over `S`, `&S`, and `&mut S`!

Here is some very rough example code I wrote when experimenting on this:
```rust
pub struct Overlay<S> {
    overlay: StagingDatabaseUpdates,
    root: S,
}

impl<S> Overlay<S> {
    fn new(root: S) -> Self {
        Self {
            root,
            overlay: Default::default(),
        }
    }
}

impl<S> Overlay<&mut S>
where
    S: CommittableSubstateDatabase + BorrowMut<S>,
{
    pub fn commit_to_root(&mut self) {
        self.root
            .borrow_mut()
            .commit(&DatabaseUpdates::from(self.overlay.clone()))
    }
}

#[test]
fn example() {
    // This works
    let db = InMemorySubstateDatabase::standard();
    let overlay = Overlay::new(db);

    // This works
    let db = InMemorySubstateDatabase::standard();
    let overlay = Overlay::new(&db);

    // This works
    let db = InMemorySubstateDatabase::standard();
    let overlay = Overlay::new(&mut db);

    // This doesn't work
    let db = InMemorySubstateDatabase::standard();
    let overlay = Overlay::new(&db);
    overlay.commit_to_root();

    // This works
    let db = InMemorySubstateDatabase::standard();
    let overlay = Overlay::new(&mut db);
    overlay.commit_to_root();
}
```

Reply:
Right! Something like this would work, exactly! 

Although slight adjustment - the overlay should be consumed when it's committed...

If it helps, I've had my own stab at this here, in case you like it / would like to merge it in: https://github.com/radixdlt/radixdlt-scrypto/pull/1730

Reply:
Thank you so much for #1730! I'm happy that `SubstateDatabaseOverlay` now works with either mutable or immutable references! I've merged the PR and just doing some minor renames of the module names.